clippy_utils/src/check_proc_macro.rs

```diff
@@ -537,6 +554,165 @@ fn ast_ty_search_pat(ty: &ast::Ty) -> (Pat, Pat) {
     }
 }
 
+/// Get the search patterns to use for the given literal
+fn token_lit_search_pat(lit: &ast::token::Lit) -> (Pat, Pat) {
+    use ast::token::LitKind;
+
+    match lit.kind {
+        LitKind::Bool => (Pat::MultiStr(&["true", "false"]), Pat::MultiStr(&["true", "false"])),
+        LitKind::Byte => (Pat::Str("b'"), Pat::Str("'")),
+        LitKind::ByteStr => (Pat::Str("b\""), Pat::Str("\"")),
+        LitKind::ByteStrRaw(0) => (Pat::Str("br\""), Pat::Str("\"")),
+        LitKind::ByteStrRaw(_) => (Pat::Str("br#\""), Pat::Str("#")),
+        LitKind::CStr => (Pat::Str("c\""), Pat::Str("\"")),
+        LitKind::CStrRaw(0) => (Pat::Str("cr\""), Pat::Str("\"")),
+        LitKind::CStrRaw(_) => (Pat::Str("cr#\""), Pat::Str("#\"")),
+        LitKind::Char => (Pat::Str("'"), Pat::Str("'")),
+        LitKind::Float | LitKind::Integer => (Pat::Sym(lit.symbol), Pat::Sym(lit.suffix.unwrap_or(lit.symbol))),
+        LitKind::Str => (Pat::Str("\""), Pat::Str("\"")),
+        LitKind::StrRaw(0) => (Pat::Str("r"), Pat::Str("\"")),
+        LitKind::StrRaw(_) => (Pat::Str("r#"), Pat::Str("#")),
+        LitKind::Err(_) => (Pat::Str(""), Pat::Str("")),
+    }
+}
+
+/// Get the search patterns to use for the given expression
+#[expect(clippy::too_many_lines, reason = "just a big `match`")]
+fn ast_expr_search_pat(e: &ast::Expr) -> (Pat, Pat) {
+    #[expect(clippy::too_many_lines, reason = "just a big `match`")]
+    fn inner(e: &ast::Expr, outer_span: Span) -> (Pat, Pat) {
+        use ast::{
+            Block, BlockCheckMode, CaptureBy, Closure, ExprKind, GenBlockKind, MatchKind, MethodCall, UnsafeSource,
+            YieldKind,
+        };
+
+        // The expression can have subexpressions in different contexts, in which case
+        // building up a search pattern from the macro expansion would lead to false positives;
+        // e.g. `return format!(..)` would be considered to be from a proc macro
+        // if we build up a pattern for the macro expansion and compare it to the invocation `format!()`.
+        // So instead we return an empty pattern such that `span_matches_pat` always returns true.
+        if !e.span.eq_ctxt(outer_span) {
+            return (Pat::Str(""), Pat::Str(""));
+        }
+
+        match &e.kind {
+            ExprKind::Underscore => (Pat::Str("_"), Pat::Str("_")),
+            ExprKind::ConstBlock(_) => (Pat::Str("const"), Pat::Str("}")),
+            ExprKind::Unary(UnOp::Deref, e) => (Pat::Str("*"), inner(e, outer_span).1),
+            ExprKind::Unary(UnOp::Not, e) => (Pat::Str("!"), inner(e, outer_span).1),
+            ExprKind::Unary(UnOp::Neg, e) => (Pat::Str("-"), inner(e, outer_span).1),
+            ExprKind::Lit(lit) => token_lit_search_pat(lit),
+            ExprKind::Paren(e) => inner(e, outer_span),
```

Comment:
Let's imagine we have `([1, 2])`. Wouldn't `ExprKind::Paren(ExprKind::Array(…))` return `("[", "]")`? Would that match?